### DISCUSSION TITLE
Using middlewares on application level

### DISCUSSION THREAD

**Discussion by serp-ya** [socketio/socket.io#3745] [Q&A] [answered]
Hi mates!

I try to create common middleware for my websocket application, but I can't to understand socket.io middlewares philosophy when I use namespaces.

What I doing:
```javascript
// server side
import { Server } from 'socket.io';

const websocket = new Server();

websocket.use((socket, next) => {
    console.log('common');
    next();
});

websocket.of('/admin').use((socket, next) => {
        console.log('admin');
        next();
    });

// client side
const socket = io.connect('/admin');
```

**Expected behavior:**
Both middlewares was executed because I use middleware on Namespace instance who was create from Server instance.

**Real behavior:**
Only Namespace's middleware was execute...

I have same issue when I try to create dynamicly namespaces with regexp:
```javascript
// server side
import { Server } from 'socket.io';

const websocket = new Server();

websocket.use((socket, next) => {
    console.log('common');
    next();
});

websocket.of(/^\/\w+$/).use((socket, next) => {
        console.log('regexp');
        next();
    });

websocket.of('/admin').use((socket, next) => {
        console.log('admin');
        next();
    });

// client side
const socket = io.connect('/admin');
```

**Expected behavior:**
Middleware on regexp namespace was executed because this is above `/admin` namespace

**Real behavior:**
Only  `/admin` Namespace middleware was execute...

**How can I to set one middleware on all namespaces?**

**NOTE: in this case I was use "standalone" initialization type, but I reproduced same behavior with different initialization types from official site**

**Comment by darrachequesne** [comment, marked as answer]
This was indeed the behavior in Socket.IO v2: even when using a namespace (`io.connect('/admin')`), you were implicitly connected to the main namespace (`/`), which is quite surprising.

This should not happen anymore with Socket.IO v3.

See also: https://socket.io/docs/v3/migrating-from-2-x-to-3-0/#No-more-implicit-connection-to-the-default-namespace

Regarding the dynamic namespace feature, the middleware registered here:

```js
websocket.of(/^\/\w+$/).use((socket, next) => {
  console.log('regexp');
  next();
});
```

is indeed registered on each child namespace. I will update the documentation about this.

**Reply by serp-ya** [reply]
Damien. hi!

Thank you for quick answer!

About my last example with dynamic namespaces - my code have next behavior:
```javascript
// server side
import { Server } from 'socket.io';

const websocket = new Server();

websocket.of(/^\/\w+$/).use((socket, next) => {
  console.log('regexp');
  next();
});

 // "admin" with single "n"
websocket.of('/admin').use((socket, next) => {
        console.log('admin');
        next();
    });

// client side
const socket = io.connect('/adminn'); // "admin" with double "n"

// Output will be 'regexp'
```

But if I connecting to "admin" with single "n" behavior was change

```javascript
// server side
import { Server } from 'socket.io';

const websocket = new Server();

websocket.of(/^\/\w+$/).use((socket, next) => {
  console.log('regexp');
  next();
});

websocket.of('/admin').use((socket, next) => {
        console.log('admin');
        next();
    });

// client side
const socket = io.connect('/admin');

// Output will be 'admin'
```

I think it looks strange because in this case I expect registration of my middleware  with regexp on each namespace, but works only one...
